### PR TITLE
 CoffeeScript -> TypeScript

### LINKED ISSUE #70: JavaScript API: Representation of empty ASTs
The representation of empty ASTs was changed in the JavaScript API when moving to the new parsing machine design. A `form` field was introduced to distinguish between a regular AST node and one that should be considered to contain no information except that parsing was successful. This adds redundant information to many AST nodes, because the user will only check for an empty AST at the top of the tree. An empty AST will never be a subtree.

1. Remove the `form` field from the `AST` class.
2. Revert to using boolean value `true` to represent a successful parse result that has no other information. This is also used in the current Python and Ruby APIs.
- OR
2. Subclass the base `AST` class with a `EmptyAST` class, similar to the Java API.

Advantage of the subclass approach is the ability to report where parsing ended from a voided starting non-terminal, in the case that it is valid to not consume all input. Disadvantage is less simplicity.

### LINKED ISSUE #72: JavaScript API: WaxeyeParser has an exposed internal method, 'match'
This change was made when moving to the new parsing machine design. Internal methods used by the `parse` method should be hidden, either with closures created in `parse`, or with an inner class as done previously.

### LINKED ISSUE #73: Ability to start parsing from any non-terminal
I suspect that people will want to be able to start parsing from an arbitrary non-terminal. The `parse` method could be overloaded in the appropriate way for each language.

### LINKED ISSUE #77: JavaScript: Rewrite in TypeScript
Easier to reason about with types, and easier to avoid mistakes / typos.

### PR DESCRIPTION
**Pull request by glebm**
Rewrites the CoffeeScript implementation in TypeScript.
The TypeScript source code is linted and auto-formatted on build.

Also makes the following changes / fixes to the parser:

* Improves performance by avoiding switching on strings.
* Splits the giant `move` method into smaller methods.
* Renames some internal types for clarity.
* Flattens the redundant MachineState (always intermediate until final) into MachineAction.
* Adds the ability to start parsing from any non-terminal.
  Fixes #73
* Hides the internal `match` method.
  Fixes #72
* Simplifies the representation for EmptyAST, removing the `from` field.
  Fixes #70
* Fixes an off-by-one in column reporting for error positions after the first line.

Fixes #77

The most important file in this PR is [waxeye.ts](https://github.com/orlandohill/waxeye/pull/78/files#diff-59eb10835816f24f5780042645b61295).

Demo: https://glebm.github.io/waxeye/demo.html

### CODE REVIEW

File: src/javascript/waxeye.ts [hunk 1]
```diff
@@ -0,0 +1,734 @@
...
+  }
+}
+
+export interface EmptyAST extends AST {
+  type: '';
+  children: never[];
+}
+
+export function EmptyAST(): EmptyAST {
+  return new AST('', []) as EmptyAST;
```
Comment (jishi9) [inline src/javascript/waxeye.ts]: Does typescript not infer the type correctly without 'as EmptyAST' ?

File: src/javascript/waxeye.ts [hunk 2]
```diff
@@ -0,0 +1,734 @@
...
+}
+
+export class ParseError {
+  constructor(
+      public pos: number, public line: number, public col: number,
+      public nt: string[], public chars: MatchError[]) {}
+
+  public toString(): string {
+    let chars;
+    chars = this.chars.map((err) => err.toGrammarString()).join(' | ');
```
Comment (jishi9) [inline src/javascript/waxeye.ts]: nit: is `err` intended to be e.g. `char` here?

File: src/javascript/package.json [hunk 3]
```diff
@@ -0,0 +1,38 @@
+{
+  "name": "waxeye",
+  "version": "0.0.1",
```
[the remark this answers is not in the document]
Comment (glebm) [reply src/javascript/package.json]: This is not a published package yet so this field does not matter

File: src/javascript/package.json [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    "compile": "node_modules/.bin/webpack",
+    "format": "node_modules/.bin/clang-format -style=Google -i",
+    "lint": "node_modules/.bin/tslint --fix --format verbose -p",
+    "test": "../../test/bin/test-javascript"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/orlandohill/waxeye.git"
+  },
+  "author": "",
```
[the remark this answers is not in the document]
Comment (glebm) [reply src/javascript/package.json]: This is not a published package yet so this field does not matter

File: src/javascript/waxeye.ts (hunk 1 above)
Comment (glebm) [reply src/javascript/waxeye.ts]: It does not. The `as` typecast will fail with any other value though (it's not like `static_cast`).

File: src/javascript/waxeye.ts (hunk 2 above)
Comment (glebm) [reply src/javascript/waxeye.ts]: No, `chars` is an array of `MatchError`s. We should probably rename it in an upcoming PR, but not int this one as `ParseError` is a public class.

File: src/javascript/package.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (glebm) [reply src/javascript/package.json]: I've removed this field and `author` because they are optional for packages marked `private`.

**Comment by orlandohill** [comment]
Looks great! Let me know, when it's ready to be merged.

**Comment by glebm** [comment]
@orlandohill Rebased, good to merge!

**Comment by orlandohill** [comment]
Great, thanks!

**Pull request merged**